Stop project generation if pre hook script fails

eliasdorneles opened the pull request:
Hey folks,

Would this be an adequate solution to address issue #464 ?

My use case is handling an invalid package name (see https://github.com/audreyr/cookiecutter-pypackage/issues/6)

This proposal will stop the project generation if the pre hook script returns with an exit status different from 0 (indicating success).

With this proposal, we'd be able to handle that case with a `pre_gen_project` hook script like:

```
import re, sys

PACKAGE_REGEX = r'^[_a-zA-Z][_a-zA-Z0-9]+$'

repo_name = '{{ cookiecutter.repo_name }}'
if not re.match(PACKAGE_REGEX, repo_name):
    print('ERROR: %s is not a valid package name!' % repo_name)
    sys.exit(1)
```

Thoughts?

eliasdorneles commented:
Okay, coverage bot, I'll add tests and improve coverage once I get some consensus that this is a good-enough approach. :)

hackebrot commented:
Hi @eliasdorneles, thank you for your contribution! :bow: 

Tbh, I'd rather raise an error instead of returning ``0``. If I am not mistaken there is no error report outside of ``generate`` that the generation failed apart from a log entry.

eliasdorneles commented:
Hi @hackebrot :)
Right, my first try was raising a custom `HookFailedException`, but then I wasn't sure where it should be caught.
Also, it would raise for all hooks (not only for `pre_gen_project`), but maybe that'd be the desired behavior?

eliasdorneles commented:
Hey, sorry for the delay -- I updated the code to use a custom exception and added cleaning up the project dir when aborting the generation, and added a little test.
Does this look better?

eliasdorneles commented:
Note: tests are broken because of issue addressed in this PR: https://github.com/audreyr/cookiecutter/pull/555

eliasdorneles commented:
Rebased to incorporate tests fix -- would appreciate some feedback. :)

I'm particularly not sure about what should be the handling the `FailedHookException` for the `post_gen_project` case -- which is why I didn't code anything for it.

hackebrot commented:
I'd say we handle errors for ``post_gen_project`` in the same fashion. :no_mouth: 

More importantly, I feel we should definitely re-raise the exception after cleaning up the filesystem instead of [returning None](https://github.com/audreyr/cookiecutter/pull/549/files#diff-b4c1da8a76ea242bf99a397dcb90b2a6R267). We can catch the ``FailedHookException`` in the callers and ``sys.exit(1)``.

cc @michaeljoseph @audreyr @pydanny

hackebrot commented on tests/test_hooks.py:
```diff
@@ -166,3 +167,16 @@ def test_run_hook(self):
+        with utils.work_in(self.repo_path):
+            with pytest.raises(exceptions.FailedHookException) as excinfo:
+                hooks.run_hook('pre_gen_project', tests_dir, {})
+            assert 'Hook script failed' in str(excinfo)
```
I think ``assert 'Hook script failed' in excinfo.value`` should work too.

eliasdorneles commented:
Thanks for reviewing @hackebrot -- I updated it with your suggestions. :)

Only thing, I think it's best not to clean up the filesystem for `post_gen_project` in the same fashion as for `pre_gen_project`. As it is a failure _after_ the project has been generated, the user wouldn't expect it to go away, and removing the project might make harder for users debug the hook scripts for their templates.

hackebrot commented:
IMHO if a ``post_gen_project`` hook fails, the generated project is likely to corrupt regardless of the actual purpose of the hook. Even if it is just debug logging messages, I expect the hook to exit without an error.

For instance, if the [rename_kv_file hook](https://github.com/hackebrot/cookiedozer/blob/master/hooks/post_gen_project.py) fails, the generated kivy app is corrupt and cannot be launched. :worried: 

So the question is why would you want to keep a broken project? I feel it is rather up to hook to use comprehensive logging messages to help users debug any error that might occur.

eliasdorneles commented:
Well, I don't have much experience using `post_gen_project` hooks, so if you say that when it fails it's likely the project to be broken, I trust you. :)

I think I was worrying unnecessarily, so lemme update this.

eliasdorneles commented:
Does this look better?

hackebrot commented on cookiecutter/generate.py:
```diff
@@ -222,6 +223,20 @@ def ensure_dir_is_templated(dirname):
+    """
+    with work_in(repo_dir):
+        try:
+            run_hook('pre_gen_project', project_dir, context)
```
I guess ``'pre_gen_project'`` is supposed to be ``hook_name``. :no_mouth:

eliasdorneles replied:
ooops, lemme fix that! :D

hackebrot commented on tests/test_hooks.py:
```diff
@@ -166,3 +167,16 @@ def test_run_hook(self):
+        hook_path = os.path.join(self.hooks_path, 'pre_gen_project.py')
+        tests_dir = os.path.join(self.repo_path, 'input{{hooks}}')
+
+        with open(hook_path, 'w') as f:
```
Uhhm, are you changing the test files at runtime? This may break other tests unintentionally... :worried:

eliasdorneles replied:
`self.repo_path` is destroyed in `teardown_method` and recreated in `setup_method` for each test:

https://github.com/audreyr/cookiecutter/pull/549/files#diff-392f380cd08f2c7cc3ac39b10cf80d95R94

hackebrot replied:
You are right :sleepy:

hackebrot commented:
Great work @eliasdorneles :+1: 

Let's see what the others think :smile_cat:

eliasdorneles commented:
Cool, thanks for reviewing @hackebrot ! :)

pydanny commented on cookiecutter/hooks.py:
```diff
@@ -104,4 +110,4 @@ def run_hook(hook_name, project_dir, context):
         logging.debug('No hooks found')
         return
-    return run_script_with_context(script, project_dir, context)
+    run_script_with_context(script, project_dir, context)
```
Why remove the ability to return a value?

eliasdorneles replied:
well, it was meant as a cleanup, `run_script_with_context` was never returning anything, so I removed the return to make clear this function is a [command not a query](https://en.wikipedia.org/wiki/Command%E2%80%93query_separation).

But I can revert the change, if you have other plans or I missed something. :)

pydanny commented:
This is a really nice feature that deserves documentation ;)

hackebrot commented:
You are dead right @pydanny! :+1:

*DocsOrDidntHappen* :smile_cat:

eliasdorneles commented:
Agreed, lemme add some docs! :)

eliasdorneles commented:
I added a note under the hook section -- does it look good?

hackebrot commented on docs/advanced_usage.rst:
```diff
@@ -34,6 +34,12 @@ hooks, as these can be run on any platform. However, if you intend for your
+.. note::
+    Make sure your hook scripts work in a robust manner.  If a hook script
+    fails (that is, if it finishes with an exit status that is not one of
+    success), the project generation will stop and the generated directory will
```
...with an exit status that is not ``0``... :grey_question:

eliasdorneles replied:
right, I was unsure if mentioning `0` would be more clear or more confusing.
I guess `0` is more clear. :)

hackebrot replied:
You could refer to the [official sys.exit docs](https://docs.python.org/2/library/sys.html#sys.exit) to be very explicit.

>  If it is an integer, zero is considered “successful termination” and any nonzero value is considered “abnormal termination” by shells and the like.

eliasdorneles replied:
Right, though the comment applies to shell script hooks too.
I'll link to it.

hackebrot commented:
I think the example that you provided in the description of the PR would be a great example for using this new feature. Let's add it to the docs :wink: 

(if @eliasdorneles and @pydanny agree :grin: )

eliasdorneles commented:
I've added an example and improved the wording -- looks good now?

hackebrot commented on docs/advanced_usage.rst:
```diff
@@ -34,6 +34,32 @@ hooks, as these can be run on any platform. However, if you intend for your
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Here is an example of script that validates a template variable
+before generating the project, to be used as ``hooks/pre_gen_project.py``::
```
I feel sorry for being this picky... :speak_no_evil: 

[PEP8 Imports](https://www.python.org/dev/peps/pep-0008/#imports) recommends the following:

```python
import re
import sys
```

Also you could enable syntax highlighting in rst via:

```
to be used as ``hooks/pre_gen_project.py``:

.. code-block:: python

    import re
```

eliasdorneles replied:
haha, I don't mind the nitpicking, keep them coming. :)

who doesn't enjoy some bikeshedding? :bike: :circus_tent: :D

eliasdorneles replied:
although... python highlighting works for me without that markup. :confused:

eliasdorneles replied:
done!

PS: wow, vim highlights sample code inline with `.. code-block::  python` :open_mouth: :notbad:

luzfcb commented on docs/advanced_usage.rst:
```diff
@@ -34,6 +34,35 @@ hooks, as these can be run on any platform. However, if you intend for your
+.. note::
+    Make sure your hook scripts work in a robust manner. If a hook script fails
+    (that is, `if it finishes with a nonzero exit status
+    <https://docs.python.org/2/library/sys.html#sys.exit>`_), the project
```
@eliasdorneles It would be nice to indicate the Python3 documentation instead of Legacy Python.

eliasdorneles replied:
:+1:

hackebrot commented:
:smile_cat: :+1:

eliasdorneles commented:
Anything else, fellows?
Do you prefer I squash the commits?

hackebrot commented:
No, you're fine! :+1: 

We are just super busy atm. Apologies for the delay. :worried: 

Going to merge this guy once we get another upvote from a contributor.

eliasdorneles commented:
gotcha!

btw, in Scrapy we find useful to rename the PR prefixing with `[MRG+1]` for these cases (PRs that were already reviewed and are only waiting for another upvote). =)

pydanny commented:
LGTM! :+1:

hackebrot commented:
Great work! :+1: :+1:

eliasdorneles commented:
Thank you! =)

hackebrot commented:
I just realized we missed sth :cry: 

We need coverage for ``cookiecutter.generate``:

```python
def _run_hook_from_repo_dir(repo_dir, hook_name, project_dir, context):
    """
    Run hook from repo directory, cleaning up project directory if hook fails
    """
    with work_in(repo_dir):
        try:
            run_hook(hook_name, project_dir, context)
        except FailedHookException:
            rmtree(project_dir)
            logging.error("Stopping generation because %s"
                          " hook script didn't exit sucessfully" % hook_name)
            raise
```

hackebrot commented:
Do you mind submitting a PR @eliasdorneles?

eliasdorneles commented:
oh, right.
sure, but it'll take me a few days, I'm on the road to Python Brasil. :)
I'll look at it after the conference.
On Nov 6, 2015 11:30 PM, "Raphael Pierzina" <notifications@github.com>
wrote:

> Do you mind submitting a PR @eliasdorneles
> <https://github.com/eliasdorneles>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/audreyr/cookiecutter/pull/549#issuecomment-154595504>.
>

hackebrot commented:
No worries! Have fun :smile_cat: 

#580
